sdc11073/mdib/devicemdib.py

```diff
@@ -495,56 +452,77 @@ def _process_transaction(self, setDeterminationTime):
-                                if oldstate:
-                                    multikey_instance.removeObjectNoLock(oldstate)
-                                multikey_instance.addObjectNoLock(newstate)
-                            break
-
-                    if corresponding_state is None:
-                        corresponding_state = self.states.descriptorHandle.getOne(descriptorContainer.handle,
-                                                                                  allowNone=True)
-                        if corresponding_state is not None:
-                            corresponding_state.descriptorContainer = descriptorContainer
-                            corresponding_state.incrementState()
-                            corresponding_state.updateDescriptorVersion()
-                    if corresponding_state is not None:
-                        updated_states.append(corresponding_state)
+                    # add state to updated_states list and to corresponding notifications input
+                    # => the state is always sent twice, a) in the description modification report and b)
+                    # in the specific state update notification.
+                    if descriptorContainer.isAlertDescriptor:
+                        update_dict = mgr.alertStateUpdates
+                    elif descriptorContainer.isComponentDescriptor:
+                        update_dict = mgr.componentStateUpdates
+                    elif descriptorContainer.isContextDescriptor:
+                        update_dict = mgr.contextStateUpdates
+                    elif descriptorContainer.isMetricDescriptor \
+                            and not descriptorContainer.isRealtimeSampleArrayMetricDescriptor:
+                        update_dict = mgr.metricStateUpdates
+                    elif descriptorContainer.isOperationalDescriptor:
+                        update_dict = mgr.operationalStateUpdates
+                    else:
+                        update_dict = mgr.rtSampleStateUpdates
+                    if descriptorContainer.isContextDescriptor:
+                        update_dict = mgr.contextStateUpdates
+                        all_states = self.contextStates.descriptorHandle.get(descriptorContainer.handle, [])
+                        for st in all_states:
+                            key = (descriptorContainer.handle, st.handle)
+                            # check if state is already present in this transaction
+                            state_update = update_dict.get(key)
+                            if state_update is not None:
+                                # the state has also been updated directly in transaction.
+                                # update descriptor version
+                                old_state, new_state = state_update
+                            else:
+                                old_state = st

```

Comment:
This case should not be legal, a descriptor update without a state update is incomplete. Only in the case of contexts this actually indicates the state was deleted.